Docs/source/usage/workflows/debugging.rst

```diff
@@ -0,0 +1,49 @@
+.. _debugging_warpx:
+
+Debugging the code
+==================
+
+Sometimes, the code does not give you the result that you are expecting.
+This can be due to a variety of reasons, from misunderstandings or changes in the :ref:`input parameters <running-cpp-parameters>`, system specific quirks, or bugs.
+You might also want to debug your code as you implement new features in WarpX during development.
+
+This section gives a step-by-step guidance on how to systematically check what might be going wrong.
+
+Debugging Workflow
+------------------
+
+Try the following steps to debug a simulation.
+
+#. Check the output text file, usually called ``output.txt``: are there warnings or errors present?
+#. On an HPC system, look for the job output and error files, usually called ``WarpX.e...`` and ``WarpX.o...``.
+   Read long messages from the top and follow potential guidance.
+#. If your simulation already created output data files:
+   Check they look reasonable before the problem occurred; are the initial conditions of the simulation as you expected?
+   Do you spot numerical artifacts or instabilities that could point to missing resolution or unexpected/incompatible numerical parameters?
+#. Did the job output files indicate a crash? Check the ``Backtrace.<mpirank>`` files for the location of the code that triggered the crash.
+   Backtraces are read from bottom (high-level) to top (most specific line that crashed).
+#. In case of a crash, Backtraces can be more detailed if you :ref:`re-compile <install-developers>` with debug flags: ``-DCMAKE_BUILD_TYPE=Debug`` (this will make the simulation slower) and rerun.
+#. If debug builds are too costly, try instead compiling with ``-DAMReX_ASSERTIONS=ON`` to activate more checks and rerun.
+#. If the problem looks like an memory violating, this could be from an invalid field or particle index access.
+   Try compiling with ``-DAMReX_BOUND_CHECK=ON`` (this will make the simulation very slow), and re-run.
+#. If the problem looks like a random memory might be used, try initializing memory with signaling Not-a-Number (NaN) values through the runtime option ``fab.init_snan = 1``.
+#. On Nvidia GPUs, if you suspect the problem might be a race condition, set the environment variable ``export CUDA_LAUNCH_BLOCKING=1`` and rerun.
```

Comment:
```suggestion
#. On Nvidia GPUs, if you suspect the problem might be a race condition due to a missing host / device synchronization, set the environment variable ``export CUDA_LAUNCH_BLOCKING=1`` and rerun.
```

Reply:
in other works it won't help with all race conditions, just ones arising from the asynchronous nature of CPU versus GPU execution.

Reply:
That's more precise, thanks!